---
title: "SpotClean adjusts for spot swapping in spatial transcriptomics data"
authors: "Zijian Ni and Christina Kendziorski"
package: SpotClean
date: "`r Sys.Date()`"
output: 
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{SpotClean}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---


```{r, echo=FALSE, results="hide", message=FALSE}
require(knitr)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
set.seed(1)
```


# Introduction

Spatial transcriptomics is a powerful and widely-used experimental method for profiling genome-wide gene expression across a tissue, which has been named [Method of the Year 2020](https://www.nature.com/articles/s41592-020-01033-y) by Nature Methods. In 10x Visium, currently the most widely used commercial protocol, fresh-frozen tissue is sectioned and placed onto a slide containing 4992 spots, with each spot containing millions of capture oligonucleotides with spatial barcodes unique to that spot. The tissue is imaged, typically via Hematoxylin and Eosin (H&E) staining. Following imaging, the tissue is permeabilized to release mRNA which binds to the capture oligonucleotides, generating a cDNA library consisting of transcripts bound by barcodes that preserve spatial information.  Data from a spatial transcriptomics experiment typically consists of the tissue image coupled with RNA-sequencing data collected from each spot. A first step in processing spatial transcriptomics data is tissue detection, where those spots on the slide containing tissue are distinguished from background spots without tissue. Unique molecular identifier (UMI) counts at each spot containing tissue are then used in downstream analyses

Ideally, a gene-specific UMI at a given spot would represent expression of that gene at that spot, and spots without tissue would show no (or few) UMIs. This is not the case in practice. Messenger RNA bleed from nearby spots causes substantial contamination of UMI counts, an artifact we refer to as spot swapping. On average, more than 30% observed expressions in a tissue spot are not originated from this spot, but from other spots contaminating it. Spot swapping confounds downstream inferences including normalization, marker gene-based annotation, differential expression and cell type decomposition.

We developed **SpotClean** to adjust for the effects of spot swapping in 10x spatial transcriptomics experiments. SpotClean is able to measure the per-spot contamination rates in observed data and decontaminate gene expression levels, thus increases the sensitivity and precision of downstream analyses. Our package `SpotClean` provides functions to load raw spatial transcriptomics data from 10x Space Ranger outputs, decontaminate the spot swapping effect, estimate contamination levels, visualize expression profiles and spot labels on the slide, and connect with other widely-used package for further analyses.


# Quick Start

## Installation

Install from Bioconductor:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SpotClean")

```

Load package after installation:

```{r}
library(SpotClean)
```

## Short Demo

Here we quickly demonstrate the general SpotClean workflow on the bundled example data. A step-by-step illustration can be found in the next section.

```{r, eval=FALSE}
# Not run

# Load example data
data(mbrain_raw)
data(mbrain_slide_info)

# Visualize raw data
mbrain_obj <- CreateSlide(count_mat = mbrain_raw, 
                          slide_info = mbrain_slide_info)
VisualizeSlide(slide_obj = mbrain_obj)
VisualizeHeatmap(mbrain_obj,rownames(mbrain_raw)[1])

# Decontaminate raw data
decont_obj <- SpotClean(mbrain_obj)

# Visualize decontaminated gene
VisualizeHeatmap(decont_obj,rownames(mbrain_raw)[1])

# Visualize the estimated per-spot contamination rate
VisualizeHeatmap(decont_obj,decont_obj@metadata$contamination_rate, 
                 logged = FALSE, legend_title = "contamination rate",
                 legend_range = c(0,1))

# (Optionally) Transform to Seurat object for downstream analyses
seurat_obj <- ConvertToSeurat(decont_obj,image_dir = "/path/to/spatial/folder")
```

## Running Speed

The computational speed is related to the size and structure of input datasets, mainly driven by the number of tissue spots. SpotClean does not require parallel computation, thus does not eat too many CPU or memory resources. As a reference, SpotClean running on a medium-size dataset (around 30,000 genes and 2,000 tissue spots) under default gene filtering takes less than 15 minutes.

## Situations you should think twice before applying SpotClean

- **Too many tissue spots**. While the observed data is a single matrix with fixed number of columns (spots), the number of unknown parameters is proportional to the number of tissue spots. In the extreme case where all spots are covered by tissue, we have more unknown parameters than observed data values. In such case the contaminated expressions are confounded with true expressions, and SpotClean estimation becomes unreliable. We recommend that the input data has at least 25% spots not occupied by tissue, so that SpotClean has enough information from background spots to estimate contamination. 

- **Low-expressed genes**. Low-expressed genes are usually of low interests and contain relatively more noises than true expression information. SpotClean by default only keeps high-expressed and high-variable genes for decontamination. It can be forced to run on manually-specified low-expressed genes, but usually does not change too much of their expressions. The fundamental trouble is actually due to the high sparsity of low-expressed genes. However, we do not filter genes by sparsity because there can be interesting genes highly concentrated in a small tissue region, and SpotClean helps removing spot swapping in that region. When the input raw data is a count matrix, high sparsity is equivalent to low expression. When input is not a count matrix, users can either adjust the expression cutoffs or manually specify genes to decontaminate.

- **Inference based on sequencing depth**. The intuition of SpotClean is to move bled-out counts to where they came from. This will significantly change the estimated sequencing depth of tissue spots after decontamination, since most estimations of sequencing depth rely on total expressions in every spot. As a result, decontamination might conflict with some sequencing depth normalization methods.

# Detailed Steps

## Load count matrix and slide information from 10x Space Ranger output

Currently, the most widely-used spatial protocol is 10x Visium. Our package provides functions to directly read 10x Space Ranger output files and generate a feature-by-barcode count matrix and slide information in R. Public 10x datasets can be found [here](https://support.10xgenomics.com/spatial-gene-expression/datasets/).

Our package contains a small subset of 10x data corresponding to the top 100 highest expressed genes of [V1_Adult_Mouse_Brain](https://support.10xgenomics.com/spatial-gene-expression/datasets/1.0.0/V1_Adult_Mouse_Brain). We now show how to process raw 10x Space Ranger output files to get this same example subset of data.

First, download and unzip the [Feature / cell matrix (raw)](https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_raw_feature_bc_matrix.tar.gz) and [Spatial imaging data](https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_spatial.tar.gz) from [V1_Adult_Mouse_Brain](https://support.10xgenomics.com/spatial-gene-expression/datasets/1.0.0/V1_Adult_Mouse_Brain). You will get a folder `raw_feature_bc_matrix` containing `barcodes.tsv.gz`, `features.tsv.gz`, `matrix.mtx.gz` and a folder `spatial` containing `aligned_fiducials.jpg`, `detected_tissue_image.jpg`, `tissue_hires_image.png`, `tissue_lowres_image.png`, `tissue_positions_list.csv`, `scalefactors_json.json`. The former folder contains the raw gene-by-spot count matrix, and the latter contains slide metadata like slide images and spot IDs and positions.

We can load these files using `Read10xRaw()` and `Read10xSlide()`, and check whether they are identical to the bundled example data:

```{r, eval=F}
# Not run

raw_mat <- Read10xRaw(count_dir = "/path/to/raw_feature_bc_matrix/")
slide_info <- Read10xSlide(tissue_csv_file="/path/to/tissue_positions_list.csv", 
                           tissue_img_file="/path/to/tissue_lowres_image.png",
                           scale_factor_file="/path/to/scalefactors_json.json")

# Compare with bundled example data
data(mbrain_raw)
data(mbrain_slide_info)
slide_info$slide$total_counts <- colSums(
    raw_mat[rownames(mbrain_raw),mbrain_slide_info$slide$barcode]
)

identical(raw_mat[rownames(mbrain_raw),], mbrain_raw)
identical(slide_info$slide, mbrain_slide_info$slide)
```

## Create the slide object

In the following, we will show the `SpotClean` workflow using the bundled example data. 

We combine count matrix and slide metadata together as one single slide object with class `SummarizedExperiment`:

```{r}
data(mbrain_raw)
data(mbrain_slide_info)

slide_obj <- CreateSlide(mbrain_raw, mbrain_slide_info)
slide_obj
```

As shown above, the raw count matrix is stored in the `raw` assay slot, and slide and image information are stored in `slide` and `grob` metadata slots. 

## Visualize the slide object

Our package provides multiple visualization functions in the 2-D slide space. Function `VisualizeSlide()` shows the input slide imaging file (if given) in R:

```{r,fig.width=5, fig.height=5}
VisualizeSlide(slide_obj)
```

Function `VisualizeLabel()` shows the spot labels. You can specify the column name of character labels in the `slide` metadata, or manually provide a vector of character labels corresponding to each spot. For example, we can plot their tissue/background labels, which has been pre-stored in the input slide information:

```{r,fig.width=5, fig.height=4}
VisualizeLabel(slide_obj,"tissue")
```


Function `VisualizeHeatmap()` draws a heatmap of values at every spot in the 2-D slide space. Similar as `VisualizeLabel()`, you can specify the column name of numerical values in the `slide` metadata, or manually provide a vector of numerical values corresponding to each spot. For example, we can plot the total UMI counts in every spot, which again has been pre-stored in the input slide information:

```{r,fig.width=5, fig.height=4}
VisualizeHeatmap(slide_obj,"total_counts")
```

You can also provide a certain gene name appearing in the raw count matrix in input slide object to `VisualizeHeatmap()`. For example, the expression of Mbp gene can be visualized:

```{r,fig.width=5, fig.height=4}
VisualizeHeatmap(slide_obj,"Mbp")
```

`VisualizeLabel()` and `VisualizeHeatmap()` both support manual label/value inputs, subsetting spots to plot, title and legend name modification. `VisualizeHeatmap()` also supports different color palettes (rainbow vs. viridis) and log scaling options. These visualization functions return `ggplot2` objects, which can be further modified by users.

## Decontaminate the data

`SpotClean()` is the main function for performing decontamination. It takes the slide object of raw data as input together with some parameters for controlling optimization and convergence, and returns a slide object with decontaminated gene expressions and other model-related parameters and statistics appending to the slide information. Detailed parameter explanations can be found by running `?SpotClean`. Here we set `maxit=10` and `candidate_radius=20` to save computation time. In practice, `SpotClean()` by default evaluates a series of candidate radius and automatically choose the best one. The default maximum number of iteration is 30, which can be extended if not converging.

```{r}
decont_obj <- SpotClean(slide_obj, maxit=10, candidate_radius = 20)
```

Check the structure of output slide object:

```{r}
decont_obj
names(decont_obj@metadata)
```

The metadata now contains more information including parameter estimations in the SpotClean model and measurements of contamination levels. 

We can visualize the Mbp gene expressions after 10 iterations of decontamination:

```{r,fig.width=5, fig.height=4}
VisualizeHeatmap(decont_obj,"Mbp")
```

## Estimate contamination levels in observed data

One important measurement people are interested in is the amount of contaminated expressions in observed expressions, i.e. expressions not originated from a tissue spot, but from other tissue spots due to spot swapping. Our model is able to estimate the proportion of contaminated expressions in each tissue spot:

```{r}
summary(decont_obj@metadata$contamination_rate)
```

This indicates around 30% expressions of a tissue spot in observed data came from spot swapping contamination, averaging across all tissue spots.

## ARC score

We also provide another subjective estimation of contamination level, called the ambient RNA contamination (ARC) score. It can be calculated using function `ARCScore()`, and is also part of the decontamination outputs. Intuitively, the ARC score is a conserved lower bound of the proportion of contamination in observed tissue spots. The ARC score can also be applied in droplet-based single-cell data to estimate ambient RNA contamination when replacing background spots with empty droplets. Details can be found by running `?ARCScore`.

```{r}
ARCScore(slide_obj)
```

This indicates at least 5% expressions in observed data came from spot swapping contamination.

## Convert to Seurat object for downstream analyses

`Seurat` is one of the most widely-used package for downstream analyses of single-cell and spatial data. We provide function `ConvertToSeurat()` to convert our slide object to Seurat spatial object. Note that Seurat requires full input of the spatial folder, which will be the `spatial` folder in the example above.

```{r, eval=F}
# Not run

seurat_obj <- ConvertToSeurat(decont_obj,image_dir = "/path/to/spatial/folder")
```

# Session Information

```{r}
sessionInfo()
```

# Citation

We appreciate it if you could cite our work when using `SpotClean`:

Ni, Z., Prasad, A., Chen, S., Arkin, L., Halberg, R., Drolet, B., Newton, M., Kendziorski, C. (2021). SpotClean adjusts for spot swapping in spatial transcriptomics data. xxx
